File: app/src/main/java/com/kieran/winnipegbus/Activities/FavouritesActivity.java
package com.kieran.winnipegbus.Activities;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemLongClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;

import com.kieran.winnipegbus.Adapters.StopListAdapter;
import com.kieran.winnipegbus.R;
import com.kieran.winnipegbusbackend.FavouriteStop;
import com.kieran.winnipegbusbackend.FavouriteStopsList;

public class FavouritesActivity extends BaseActivity implements AdapterView.OnItemClickListener, OnItemLongClickListener {
    private StopListAdapter adapter;

    @Override
    protected void onRestart() {
        super.onRestart();
        FavouriteStopsList.sort(getSortPreference());
        reloadList();
    }

    @Override
    protected void onResume() {
        super.onResume();
        FavouriteStopsList.sort(getSortPreference());
        reloadList();
    }

    private void reloadList() {
        FavouriteStopsList.isLoadNeeded = true;
        getFavouritesList();
        adapter.notifyDataSetChanged();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        adViewResId = R.id.stopsListAdView;

        setContentView(R.layout.activity_favourite_stops);

        ListView listView = (ListView) findViewById(R.id.stops_listView);

        initializeAdsIfEnabled();
        getFavouritesList();

        listView.setOnItemClickListener(this);
        listView.setOnItemLongClickListener(this);

        adapter = new StopListAdapter(this, R.layout.listview_stops_row);
        listView.setAdapter(adapter);
    }

    private void getFavouritesList() {
        FavouriteStopsList.loadFavourites();
        StopListAdapter.sortPreference =  getSortPreference();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_favourites, menu);
        return true;
    }

    private void openStopTimesAndUse(FavouriteStop favouriteStop) {
        favouriteStop.use();
        FavouriteStopsList.saveFavouriteStops();

        openStopTimes(favouriteStop);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        openStopTimesAndUse(adapter.getItem(position));
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, final int position, long id) {
        final Context context = this;
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);

        alertDialog.setMessage("Edit this Favourite?");
        alertDialog.setPositiveButton("Delete", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialogInterface, int which) {
                FavouriteStopsList.remove(adapter.getItem(position).getNumber());
                reloadList();
            }
        });

        alertDialog.setNeutralButton("Rename", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialogInterface, int which) {
                AlertDialog.Builder renameDialog = new AlertDialog.Builder(context);
                final EditText editText = new EditText(context);
                final FavouriteStop favouriteStop = FavouriteStopsList.get(position);
                editText.setText(favouriteStop.getDisplayName());
                renameDialog.setView(editText);

                renameDialog.setNeutralButton("Default", null);

                renameDialog.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        FavouriteStopsList.get(position).setAlias(editText.getText().toString());
                        FavouriteStopsList.saveFavouriteStops();
                        reloadList();
                    }
                });
                renameDialog.setNegativeButton("Cancel", null);

                Button button = renameDialog.show().getButton(DialogInterface.BUTTON_NEUTRAL);
                button.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        editText.setText(favouriteStop.getName());
                    }
                });
            }
        });

        alertDialog.setNegativeButton("Cancel", null);
        alertDialog.create().show();

        return true;
    }
}

File: app/src/main/java/com/kieran/winnipegbus/activities/FavouritesActivity.kt
package com.kieran.winnipegbus.activities

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemLongClickListener
import android.widget.EditText
import android.widget.ListView
import com.kieran.winnipegbus.R
import com.kieran.winnipegbus.adapters.StopListAdapter
import com.kieran.winnipegbusbackend.common.FavouriteStop
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class FavouritesActivity : BaseActivity(), AdapterView.OnItemClickListener, OnItemLongClickListener {
    private var adapter: StopListAdapter? = null
    private val favouriteStops = ArrayList<FavouriteStop>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adViewResId = R.id.stopsListAdView

        setContentView(R.layout.activity_favourite_stops)

        val listView = findViewById<ListView>(R.id.stops_listView)

        initializeAdsIfEnabled()

        listView.onItemClickListener = this
        listView.onItemLongClickListener = this

        adapter = StopListAdapter(this, R.layout.listview_stops_row, favouriteStops)
        listView.adapter = adapter
    }

    override fun onRestart() {
        super.onRestart()
        reloadList()
    }

    override fun onResume() {
        super.onResume()
        reloadList()
    }

    private fun reloadList() {
        GlobalScope.launch(Dispatchers.IO) {
            val stops = favouritesService.getAll(sortPreference)

            runOnUiThread {
                favouriteStops.clear()
                favouriteStops.addAll(stops)
                adapter?.notifyDataSetChanged()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_favourites, menu)
        return true
    }

    private fun openStopTimesAndUse(favouriteStop: FavouriteStop) {
        favouriteStop.use()
        favouritesService.update(favouriteStop)

        openStopTimes(favouriteStop)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        openStopTimesAndUse(adapter!!.getItem(position) as FavouriteStop)
    }

    override fun onItemLongClick(parent: AdapterView<*>, view: View, position: Int, id: Long): Boolean {
        val context = this
        val alertDialog = AlertDialog.Builder(this)

        alertDialog.setMessage(R.string.edit_favourite_dialog_title)
        alertDialog.setPositiveButton(R.string.delete) { _, _ ->
            favouritesService.delete((adapter!!.getItem(position)!!.id))
            reloadList()
        }

        alertDialog.setNeutralButton(R.string.rename) { _, _ ->
            val renameDialog = AlertDialog.Builder(context)
            val editText = EditText(context)
            val favouriteStop = favouriteStops[position]
            editText.setText(favouriteStop.displayName)
            renameDialog.setView(editText)

            renameDialog.setNeutralButton(R.string.default_label, null)

            renameDialog.setPositiveButton(R.string.ok) { _, _ ->
                favouriteStop.alias = editText.text.toString()
                favouritesService.update(favouriteStop)
                reloadList()
            }
            renameDialog.setNegativeButton(R.string.cancel, null)

            val button = renameDialog.show().getButton(DialogInterface.BUTTON_NEUTRAL)
            button.setOnClickListener { editText.setText(favouriteStop.name) }
        }

        alertDialog.setNegativeButton(R.string.cancel, null)
        alertDialog.create().show()

        return true
    }
}
